src/OpalCompiler-Core/SelfVariable.class.st

```diff
@@ -35,3 +35,14 @@ SelfVariable >> isSelfVariable [
 SelfVariable >> readInContext: aContext [
 	^aContext receiver
 ]
+
+{ #category : #queries }
+SelfVariable >> usingMethods [
+	"as super sends are doing a pushSelf, too, we need to still check the AST level sometimes"
+
+	^ SystemNavigation new allMethods select: [ :method | 
+		  method readsSelf and: [ 
+			  method sendsToSuper not or: [ 
+				  method ast variableNodes anySatisfy: [ :varNode | 
```

Comment:
I think we do not need ast traversal here. It's enough to just detect a self send.
We can also add #sendsToSelf in addition to #sendsToSuper.

Reply:
Notice I introduced singletons for these variables. So we have very simple queries like:
```Smalltalk 
SuperVariable instance usingMethods.
ThisContextVariable instance usingMethods
```

Reply:
the problem is that "super something" actually is "pushSelf doSuperSend", so scanning for self access gives us all the super, too.

Reply:
self can be referenced standalone in addition (e.g. self something ifTrue: [^self].)